### PR TITLE
EIP-2981: ERC-721 Royalty standard - Standardized means of accepting royalties for NFT marketplaces across the ecosystem

### PR DESCRIPTION
**Pull request by VexyCats**
full EIP text here : https://github.com/VexyCats/EIPs/blob/master/EIPS/eip-2981.md
Conversation here : #2907 https://github.com/ethereum/EIPs/issues/2907

### CODE REVIEW

**Comment by VexyCats** [comment]
Hey @MicahZoltu - what am I missing that made this CI check fail? Anything look out of place to you? Says its failing on my last commit which was just changing some spacing on a code bracket

File: EIPS/EIP-1776.md
Comment (MicahZoltu) [inline EIPS/EIP-1776.md:6]:
```suggestion
Draft
```

**Comment by MicahZoltu** [comment]
Also, you have several typos that the bot doesn't like:
```
./EIPS/EIP-1776.md:15: publically ==> publicly
./EIPS/EIP-1776.md:23: flexibilty ==> flexibility
./EIPS/EIP-1776.md:25: transfering ==> transferring
./EIPS/EIP-1776.md:34: recieving ==> receiving
./EIPS/EIP-1776.md:61: transfered ==> transferred
./EIPS/EIP-1776.md:76: recieve ==> receive
./EIPS/EIP-1776.md:78: transfered ==> transferred
./EIPS/EIP-1776.md:158: Transfering ==> Transferring
```

**Comment by VexyCats** [comment]
Hey thanks! No spell check in my IDE ;D 

Will go through and check everything for typos on an external program and resubmit! 

(Didn't know typos affected the PR acceptance)

**Review by MicahZoltu — APPROVED** [review]
I left a bunch of feedback which I encourage you to consider, but for getting this merged as a Draft all that is required is that the EIP is structurally sound (meaning you have the right sections and the frontmatter is setup correctly).

Since I left some inline suggestions in here, I'll wait for you to indicate you want me to merge since once I merge you can't do one-click apply of the inline suggestions anymore.

File: EIPS/EIP-1776.md
Comment (wighawag) [inline EIPS/EIP-1776.md:2]: EIP-1776 is already taken : https://github.com/ethereum/EIPs/issues/1776

File: EIPS/EIP-1776.md
Comment (VexyCats) [reply EIPS/EIP-1776.md:36]: I choose fixed point with scaling to 100 as all the other marketplaces with royalties do not let you go below 1%, with the average royalty percentage being 1-10%. 

I'm not opposed to making it go lower. I'm not sure which would be best, a scaling factor of 10,000, or scaled to 10^18. 

Most users would not need/select a percentage of 0.00005% though - as an NFT purchase is a onetime transaction and not like a general fee on all transactions on something like say uniswap. 

Continuing to go through these and replying. Thank you for the feedback Michah!

File: EIPS/EIP-1776.md
Comment (VexyCats) [reply EIPS/EIP-1776.md:19]: Will adjust this. Thank you.

File: EIPS/EIP-1776.md
Comment (VexyCats) [reply EIPS/EIP-1776.md:10]: The hard dependency is on 165 for the magic value to be returned that the implementation of royalties is in the NFTs.... so yes, you are right, you actually don't need to implement 721. Will adjust accordingly.

File: EIPS/EIP-1776.md
Comment (VexyCats) [reply EIPS/EIP-1776.md:48]: Yes. I had it originally but 721 requires 165, so by defacto would be required. Although as per your last comment, 721 _isn't actually_ required.  Will adjust.

File: EIPS/EIP-1776.md
Comment (VexyCats) [reply EIPS/EIP-1776.md:38]: Both implement the standard. NFTs for setting and returning the royalty values, and marketplaces for upholding the royalty payment on a transaction. 

Will remove the last sentence. Thank you.

File: EIPS/EIP-1776.md
Comment (VexyCats) [reply EIPS/EIP-1776.md:59]: Hmm. Yes this would be better suited as a stand alone `contract` rather than `interface`.

File: EIPS/EIP-1776.md
Comment (VexyCats) [reply EIPS/EIP-1776.md:73]: Yes `royaltiesReceived` must be called after a transfer of funds. I will add this `MUST`. 

Also will change this from an interface into a contract, as its better suited as a contract. 

TokenId can be included. Will make the changes.

File: EIPS/EIP-1776.md
Comment (VexyCats) [reply EIPS/EIP-1776.md:88]: Roger. Will adjust. This is my code of implementation - will remove this and clean it up to remove implementation leaving only specifics for the standard.

File: EIPS/EIP-1776.md
Comment (VexyCats) [reply EIPS/EIP-1776.md:94]: No - I considered weighing the royalty per NFT - and there is a few pros and a few cons. 

Overall the main reason to go towards a royalty per contract (all the NFTs in this one contract) was due to gas costs during minting. 

Setting a royalty value per NFT works - but prohibits the ability to mint multiple NFTs in a single transaction - by increasing the state per NFT. I've redone the 721 standard for batch minting and can achieve upwards of 2,500 NFTs in a single transaction, but if royalties are set per NFT, it would be less than 40 per NFT. Same goes for setting a value like URI instead of using a chronological URI pattern and baseURI. 

I see the value prop of having a royalty per NFT, but there are downsides for the creator. If the creator wanted to set royalties for all their NFTs - in the long run of minting each NFT they would be paying more than if they were able to set the base royalty amount. 

Thinking about it further - you could handle it the same way as a baseURI and (this is purely implementation) return the base royalty amount for all NFTs, or individual amount if its set. 

Therefore the function could take a tokenId argument, and the returned value only needs to be a uint256, regardless if its a baseURI or not. 

Okay - I've convinced myself lol. Will adjust changed and allow for tokenId input. Makes sense as it offers more flexibility, albeit costing more gas during implementations.

File: EIPS/EIP-1776.md
Comment (VexyCats) [reply EIPS/EIP-1776.md:48]: Yes thank you. As mentioned above it will be converted to a contract as this was my implementation and interface probably shouldn't be included for the standard

File: EIPS/EIP-1776.md
Comment (VexyCats) [reply EIPS/EIP-1776.md:101]: Will adjust. Thank you

File: EIPS/EIP-1776.md
Comment (VexyCats) [reply EIPS/EIP-1776.md]: Roger. What would you like to see for the example section, if not this?

File: EIPS/EIP-1776.md
Comment (VexyCats) [reply EIPS/EIP-1776.md:174]: Thanks, will adjust it accordingly.

File: EIPS/EIP-1776.md
Comment (VexyCats) [reply EIPS/EIP-1776.md:24]: Thanks will do.

File: EIPS/EIP-1776.md
Comment (VexyCats) [reply EIPS/EIP-1776.md:2]: Thanks, dang I searched and thought 1776 was free, yours is not in EIPs or a PR so that is why I did not see it, sorry about that. Are you going to do a PR for it? 

Will adjust over to a new value for the EIP

File: EIPS/EIP-1776.md
Comment (MicahZoltu) [reply EIPS/EIP-1776.md:94]: I think whether royalties are contract wide or not is an implementation detail.  One could imagine a token contract where all tokens have the same royalty, in which case no matter what tokenID you pass in you get the same result, but updating it merely requires updating a single variable.  Another token may have a default fee that can be overridden by individual tokens, and a third may have a fee rate per token.

File: EIPS/EIP-1776.md
Comment (MicahZoltu) [reply EIPS/EIP-1776.md:48]: > it will be converted to a contract as this was my implementation and interface probably shouldn't be included for the standard

Do you mean the other way around?  Interface standards should never have implementation details in the specification, though sometimes a reference implementation may be provided in the `## Implementation` section.

File: EIPS/EIP-1776.md
Comment (MicahZoltu) [reply EIPS/EIP-1776.md]: I'm generally not a fan of an examples subsection at all.  I'm a big fan of the specification just being a specification, and people can use other resources for examples, tutorials, how-tos, deep dives, etc.  I prefer the spec to be generally pretty short and to the point, describing exactly what is necessary to be spec compliant and nothing else.

File: EIPS/EIP-1776.md
Comment (MicahZoltu) [inline EIPS/EIP-1776.md:2]:
```suggestion
eip: 2981
```
Good catch @wighawag.  Please use 2981 for this EIP, and rename the file to `eip-2981.md` (note the casing).

File: EIPS/EIP-1776.md
Comment (MicahZoltu) [inline EIPS/EIP-1776.md:162]:
```suggestion
   IERC2981(_tokenAddress).royaltiesReceived(_recipient, _buyer, amount);
```

File: EIPS/EIP-1776.md
Comment (wighawag) [reply EIPS/EIP-1776.md:2]: yes, still planning to work on it

File: EIPS/EIP-1776.md
Comment (VexyCats) [reply EIPS/EIP-1776.md]: I used 10,000 as the scaling factor as I don't see any reason someone would want to go below that for an NFT royalty fee. 

Im unsure if its a good move, as for adopting the standard a marketplace really needs to make sure its math is correct and it uses the 10000 scale factor, otherwise it might  have a bad time..... I'd like an easier solution where it requires less math/attention to detail for implementation, but alas I don't know of what would be better

**Comment by VexyCats** [comment]
@MicahZoltu I think I've made all the changes - please feel free to merge the PR, unless you have any other feedback. As we get more discussions and thoughts on it can always propose changes 

Thank you for your feedback so far!

File: EIPS/EIP-1776.md
Comment (MicahZoltu) [reply EIPS/EIP-1776.md]: The advantage of `10^18` is that it is *very* standard across Ethereum, so it is more likely that tools will "just work" with that then with some custom scaling factor.  The advantage of `100` is that it is easy to just describe the value as a "percentage".  Personally, I would just do `10^18` I think because I believe that 100 is way too limiting (I can imagine people wanting sub percentage royalties) and I think standardization is valuable.  You aren't really saving anything by using 10,000 as all numbers in Solidity use the same number of bytes of storage (unless you are bit packing, which I don't think you are or should).

**Review by MicahZoltu — APPROVED** [review]

**Pull request merged**
